### PR TITLE
feat: routes and components generation GENC-68 

### PR DESCRIPTION
**Pull request by skawian**
[Review Guidance](https://www.notion.so/genesisglobal/Platform-Code-Review-Process-ace93ba760cc4563b0dfb712e2a88d8a)

📓 &nbsp; **Related JIRA**

[GENC-68](https://genesisglobal.atlassian.net/browse/GENC-68)

🤔 &nbsp; **What does this PR do?**

- routes and components generation using templates

🚀 &nbsp; **Where should the reviewer start?**

Add file / directory pointers.

📑 &nbsp; **How should this be tested?**

Providing config with prompts:
```
npx -y @genesislcap/genx@latest init prtest -x --ref ms/GENC-68-generating-components --apiHost 'wss://public-foundation.genesislab.global/gwf/'
```

Providing config with parameter:

```
npx -y @genesislcap/genx@latest init prtest -x --ref ms/GENC-68-generating-components --routes '[{"name":"home","tiles":[{"title":"Entity manager tile","type":"entity-manager","config":{"resourceName":"ALL_COUNTERPARTYS","title":"Counterparty Management","updateEvent":"EVENT_COUNTERPARTY_MODIFY","deleteEvent":"EVENT_COUNTERPARTY_DELETE","createEvent":"EVENT_COUNTERPARTY_INSERT"}},{"title":"Form tile","type":"smart-form","config":{"resourceName":"EVENT_TRADE_INSERT"}}]},{"name":"page-2","tiles":[{"title":"Grid Tile","type":"grid-pro","config":{"resourceName":"ALL_POSITIONS"}},{"title":"Charts Tile 1","type":"chart","config":{"type":"line","resourceName":"ALL_POSITIONS","xField":"INSTRUMENT_NAME","yField":"VALUE"}},{"title":"Charts Tile 2","type":"chart","config":{"type":"pie","resourceName":"ALL_POSITIONS","xField":"INSTRUMENT_NAME","yField":"VALUE"}},{"title":"Charts Tile 3","type":"chart","config":{"type":"column","resourceName":"ALL_POSITIONS","xField":"INSTRUMENT_ID","yField":"VALUE"}}]}]' --apiHost 'wss://public-foundation.genesislab.global/gwf/'
```

No routes provided (sets default):
```
npx -y @genesislcap/genx@latest init prtest -x --ref ms/GENC-68-generating-components --apiHost 'wss://public-foundation.genesislab.global/gwf/'
```

Follow the video for testing instructions

https://github.com/genesiscommunitysuccess/blank-app-seed/assets/241069/d380f1e4-2ad5-4d1c-bfc9-8ef37daaba79

sample json config for pages:

```
[{"name":"home","tiles":[{"title":"Entity manager tile","type":"entity-manager","config":{"resourceName":"ALL_COUNTERPARTYS","title":"Counterparty Management","updateEvent":"EVENT_COUNTERPARTY_MODIFY","deleteEvent":"EVENT_COUNTERPARTY_DELETE","createEvent":"EVENT_COUNTERPARTY_INSERT"}},{"title":"Form tile","type":"smart-form","config":{"resourceName":"EVENT_TRADE_INSERT"}}]},{"name":"page-2","tiles":[{"title":"Grid Tile","type":"grid-pro","config":{"resourceName":"ALL_POSITIONS"}},{"title":"Charts Tile 1","type":"chart","config":{"type":"line","resourceName":"ALL_POSITIONS","xField":"INSTRUMENT_NAME","yField":"VALUE"}},{"title":"Charts Tile 2","type":"chart","config":{"type":"pie","resourceName":"ALL_POSITIONS","xField":"INSTRUMENT_NAME","yField":"VALUE"}},{"title":"Charts Tile 3","type":"chart","config":{"type":"column","resourceName":"ALL_POSITIONS","xField":"INSTRUMENT_ID","yField":"VALUE"}}]}]
```

✅ &nbsp; **Checklist**

- [x] I have tested my changes.
- [x] I have added tests for my changes.
- [x] I have updated the project documentation to reflect my changes.

### CODE REVIEW

File: .genx/prompts/ui.js [hunk 1]
```diff
@@ -10,13 +10,23 @@ module.exports = async (inquirer, prevAns = {}) => {
...
-      default: prevAns.routes || 'home',
+      default: `[{name: 'home'}]`,
     },
   ])
 
+  let routesParsed = [{name: 'home'}];
+  if (routes) {
+    try {
+      routesParsed = JSON.parse(routes);
+      console.log(routesParsed)
```
Comment (cistov) [inline .genx/prompts/ui.js:23]: We can remove this log - we'll know it was valid if there's no error

Comment (skawian) [reply .genx/prompts/ui.js]: yeah, looks like leftover from debugging here

**Comment by skawian** [comment]
> Could be on my end but got this when running the test command:
> 
> Error parsing JSON, will use default routes config: SyntaxError: Expected property name or '}' in JSON at position 2 at JSON.parse () at module.exports (/home/bruno/genesis/prtest/.genx/prompts/ui.js:22:27) at async module.exports (/home/bruno/genesis/prtest/.genx/prompts.js:16:20) ✔ Project successfully created in 3411ms.
> 
> But happy with code changes!

can you try again? I modified the instructions

File: .genx/prompts/ui.js [hunk 2]
```diff
@@ -10,13 +10,22 @@ module.exports = async (inquirer, prevAns = {}) => {
     {
       name: 'routes',
       type: 'input',
-      message: 'Pages in comma separated format e.g. home,dashboard,admin',
+      message: 'Pages config in json format',
       when: !prevAns.routes,
-      default: prevAns.routes || 'home',
+      default: `[{name: 'home'}]`,
```
Comment (cistov) [inline .genx/prompts/ui.js:15]: This default is redundant now with the code below

File: .genx/configure.js [hunk 3]
```diff
@@ -18,14 +18,21 @@ module.exports = async (data, utils) => {
   data.localGenId = data.appName.toUpperCase().replace("-", "_");
   data.applicationVersionWeb = data.applicationVersion.split('-').shift();
   data.versions = versions;
+
+  utils.registerPartial('smart-form', path.resolve(data.directory, '.genx/templates/form.hbs'))
+  utils.registerPartial('chart', path.resolve(data.directory, '.genx/templates/chart.hbs'))
+  utils.registerPartial('entity-manager', path.resolve(data.directory, '.genx/templates/entityManager.hbs'))
+  utils.registerPartial('grid-pro', path.resolve(data.directory, '.genx/templates/grid.hbs'))
+  // utils.registerPartial('form', ``)
```
Comment (cistov) [inline .genx/configure.js:26]: This comment can go

File: .genx/prompts/ui.js [hunk 4]
```diff
@@ -10,13 +10,22 @@ module.exports = async (inquirer, prevAns = {}) => {
...
     },
   ])
 
+  let routesParsed = [{name: 'home'}];
+  if (routes) {
+    try {
+      routesParsed = JSON.parse(routes);
+    } catch (error) {
+      console.error("Error parsing JSON, will use default routes config:", error);
```
Comment (cistov) [inline .genx/prompts/ui.js:24]: Not passing `routes` results in below, which looks like there's something wrong:

```
Error parsing JSON, will use default routes config: SyntaxError: Expected property name or '}' in JSON at position 2
    at JSON.parse (<anonymous>)
    at module.exports (/Users/artur.cistov/code/prtest2/.genx/prompts/ui.js:22:27)
    at async module.exports (/Users/artur.cistov/code/prtest2/.genx/prompts.js:16:20)
```

It should be like this: you didn't specify routes, we silently we switch to the default.
If you did specify, but it's invalid JSON, we should print just `error.message` (without the stacktrace) - something like this:

```
console.error("Error parsing `routes` parameter as JSON:", error.message);
console.log("Falling back to the default routes value");
```

File: .genx/templates/entityManager.hbs [hunk 5]
```diff
@@ -0,0 +1,18 @@
+<entity-management
+    {{#if config.updateEvent}}
```
Comment (cistov) [inline .genx/templates/entityManager.hbs:2]: should be `{{#if config.title}}` ?

File: .genx/templates/grid.hbs [hunk 6]
```diff
@@ -0,0 +1,9 @@
+<zero-grid-pro>
+  <grid-pro-genesis-datasource
+    resource-name="{{config.resourceName}}"
+    {{#if config.snapshot}}
+    isSnapshot="{{config.resourceName}}"
```
Comment (cistov) [inline .genx/templates/grid.hbs:5]: `isSnapshot="{{config.snapshot}}"`?

**Review by cistov — APPROVED** [review]

**Pull request merged**